relay-redis/src/real.rs

```diff
@@ -382,3 +392,160 @@ pub struct Stats {
+    type Connection = MultiplexedConnection;
+    type Error = redis::RedisError;
+
+    async fn connect(&self) -> Result<Self::Connection, Self::Error> {
+        self.client
+            .get_multiplexed_async_connection_with_config(&self.config)
+            .await
+    }
+
+    async fn is_valid(&self, conn: &mut Self::Connection) -> Result<(), Self::Error> {
+        check_redis_is_valid(conn).await
+    }
+
+    fn has_broken(&self, _: &mut Self::Connection) -> bool {
+        false
+    }
+}
+
+/// [`bb8::ManageConnection`] for an async redis cluster.
+pub struct RedisClusterConnectionManager {
+    client: ClusterClient,
+}
+
+impl RedisClusterConnectionManager {
+    /// Creates a new [`bb8::ManageConnection`] with the provided [`ClusterClient`]
+    pub fn new(client: ClusterClient) -> Self {
+        Self { client }
+    }
+}
+
+#[axum::async_trait]
+impl bb8::ManageConnection for RedisClusterConnectionManager {
+    type Connection = ClusterConnection;
+    type Error = redis::RedisError;
+
+    async fn connect(&self) -> Result<Self::Connection, Self::Error> {
+        self.client.get_async_connection().await
+    }
+
+    async fn is_valid(&self, conn: &mut Self::Connection) -> Result<(), Self::Error> {
+        check_redis_is_valid(conn).await
+    }
+
+    fn has_broken(&self, _: &mut Self::Connection) -> bool {
+        false
+    }
+}
+
+async fn check_redis_is_valid<C: redis::aio::ConnectionLike>(conn: &mut C) -> RedisResult<()> {
+    let pong: String = redis::cmd("PING").query_async(conn).await?;
+    match pong.as_str() {
+        "PONG" => Ok(()),
+        _ => Err((ErrorKind::ResponseError, "ping request").into()),
+    }
+}
+
+/// An abstraction over Single and Cluster redis connections.
+#[derive(Debug, Clone)]
+pub enum AsyncRedisPool {
+    /// Represents a connection pool to a clustered redis instance.
+    Cluster(bb8::Pool<RedisClusterConnectionManager>, RedisConfigOptions),
+    /// Represents a connection pool to a single redis server.
+    Single(
+        bb8::Pool<RedisConnectionManagerWithConfig>,
+        RedisConfigOptions,
+    ),
+}
+
+impl AsyncRedisPool {
+    /// Takes a command and executes it on redis.
+    pub async fn query_async<T: FromRedisValue>(&self, cmd: Cmd) -> Result<T, RedisError> {
+        match self {
+            Self::Cluster(pool, ..) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                cmd.query_async(&mut *conn).await.map_err(RedisError::Redis)
+            }
+            Self::Single(pool, ..) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                cmd.query_async(&mut *conn).await.map_err(RedisError::Redis)
+            }
+        }
+    }
+
+    /// Creates a new cluster based [`AsyncRedisPool`].
+    /// It will set the `response_timeout` to the provided `read_timeout`.
+    /// `write_timeout` is not used at all in the async pool
```

Comment:
```suggestion
    /// Creates a new cluster based [`AsyncRedisPool`].
    ///
    /// It will set the `response_timeout` to the provided `read_timeout`,
    /// `write_timeout` is not used at all in the async pool.
```